Fixed #28300 -- Allowed GDALRaster objects to use the vsimem filesystem.

yellowcap opened the pull request:
https://code.djangoproject.com/ticket/28300

timgraham commented:
buildbot, test on xenial.

timgraham commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -121,7 +121,7 @@ def test_memory_based_raster_creation(self):
         self.assertEqual(result, list(range(256)))
 
-    def test_file_based_raster_creation(self):
+    def test_file_path_based_raster_creation(self):
```
Should this rename be in a separate commit?

timgraham commented on docs/releases/2.0.txt:
```diff
@@ -86,6 +86,10 @@ Minor features
 
+* Allowed creating :class:`~django.contrib.gis.gdal.GDALRaster` objects in
+  GDAL's internal virtual filesystem. Rasters can now be created from and
+  converted to binary data in memory directly.
```
Link to the new documentation in the release note.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
+    ...     "scale": [100, -100],
+    ...     "bands": [{"data": range(36), "nodata_value": 99}]
+    ... })
+    >>> return HttpResponse(rast.vsi_buffer, 'image/tiff')
```
A return statement doesn't make sense in a shell session. Also, HttpResponse import is missing.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
+
+    # Create a new GTiff raster in a view and return it as a file.
+    >>> rst = GDALRaster({
+    ...     "name": "/vsimem/temporarymemfile",
```
Use single quotes

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
+
+    # Read a raster as file object from a remote source.
+    >>> import urllib
+    >>> dat = urllib.urlopen('http://example.com/raster.tif').read()
```
urllib.urlopen looks like Python 2

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1105,15 +1105,25 @@ blue.
+    and the new raster is always created in write mode.
+
+    The first parameter can take three forms: a string representing a file path, a
+    dictionary with values defining a new raster, or a bytes object representing a raster
```
Line length is exceed 79 characters in the new text.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
+    ...     "bands": [{"data": range(36), "nodata_value": 99}]
+    ... })
+    >>> return HttpResponse(rast.vsi_buffer, 'image/tiff')
+
```
1 blank only

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
+    # The name starts with /vsimem/, indicating that the raster lives in the
+    # virtual filesystem.
+    >>> rst.name
+    /vsimem/da300bdb-129d-49a8-b336-e410a9428dad
```
Should that be a quoted string?

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
+dictionary representation, and provide a ``name`` argument that starts with
+``/vsimem/`` (for detail on the dictionary representation see
+:ref:`gdal-raster-ds-input`). For virtual file based rasters, the
+:attr:`~GDALRaster.vsi_buffer` will return the ``bytes`` representation of the
```
will return -> returns

timgraham commented on django/contrib/gis/gdal/raster/const.py:
```diff
@@ -43,3 +43,13 @@
+
+# Flag for managing ownership over buffer when creating VSI in
+# memory files.
+VSI_TAKE_BUFFER_OWNERSHIP = False
```
Are these boolean constants just for readability or are these settings meant to be monkeypatched for alternate behavior?

yellowcap replied:
These are meant for readability only.

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -66,13 +73,39 @@ def __init__(self, ds_input, write=False):
+            self._write = 1
+            # Get size of buffer.
+            size = sys.getsizeof(ds_input)
+            # Pass data to ctypes, keep reference to the ctypes object.
```
, keeping a reference to the cyptes object so that the vsimem file...

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -151,6 +184,13 @@ def __init__(self, ds_input, write=False):
         else:
             raise GDALException('Invalid data source input type: "{}".'.format(type(ds_input)))
 
+    def __del__(self):
```
A task in https://code.djangoproject.com/ticket/23919 is "Evaluate replacement of custom `__del__` methods by  [weakref.finalize()](https://docs.python.org/3/library/weakref.html#weakref.finalize)". I wonder if you have any input about that?

yellowcap replied:
Unfortunately I can't give input on that. I have never heard of `weakref` before, and I don't know enough about pythons gc etc to have an informed opinion.

yellowcap commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -66,13 +73,39 @@ def __init__(self, ds_input, write=False):
-            if not os.path.exists(ds_input):
-                raise GDALException('Unable to read raster source input "{}"'.format(ds_input))
+            if not ds_input.startswith(VSI_FILESYSTEM_BASEPATH) and not os.path.exists(ds_input):
+                raise GDALException('Opening rasters from vsi file paths is not allowed.')
```
Thanks for spotting this, the error message like this does not make sense. I had a previous version where direct opening of vsi paths was not allowed. But it should work like files, its the user's responsibility when opening the same vsi file twice.

I actually think this check is now not necessary anymore, this "file exists" check was created when only str input was allowed.  The try/except error message should be explicit enough. So I propose to remove those two lines entirely. What do you think?

yellowcap commented:
- Added a test for `vsi_buffer` on non-vsi rasters.
- Removed the additional path check from the `str` based initiation of GDALRasters.
- Squashed the edits commit and added acknowledgements in the commit message. Hope that is ok.

yellowcap commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -173,6 +213,25 @@ def _flush(self):
     @property
+    def vsi_buffer(self):
+        if not self.is_vsi_based:
+            return None
```
Added test for it [here](https://github.com/django/django/pull/8634/files#diff-469b4a6b934be489c22f3cf20c6e0375R218).
